### ISSUE TITLE
SuggestionList. Array of suggestions API

### ISSUE THREAD

**Issue by GeorgeTaveras1231** [xogroup/union#249]
Does it really make sense to support a suggestion array prop on the `SuggestionList` component?

```jsx
const suggestions = [{id: 1, description : "suggestion label"}];

<SuggestionsList suggestions={suggestions} />
```

My understanding is that it is intended to save the consumer the need to map over the data from an API and convert each entry to a `<SuggestionListItem />`. 
My concern is that the constraints of the array items is still really strict (must have `id` and `description`), which more than likely means the consumer still needs to map over their data in order to make it compatible with the `SuggestionList`. Unless there is something im missing, this seems to me like it is trying to solve a problem yet not really solving it.

**Comment by marioharper** [comment]
What about something like this
```
<SuggestionList 
  suggestions={[{label: 'suh'}, {label: 'dude'}]} 
  renderListItem={(suggestion) => (
    <div>{suggestion.label}</div>
  )} 
/>
```

**Comment by marioharper** [comment]
sort of like this http://ericgio.github.io/react-bootstrap-typeahead/

**Comment by GeorgeTaveras1231** [comment]
@marioharper to me this seems like the `SuggestionList` component wants to have too much control without any benefit. I don't see a need for the `SuggestionList` component to control the mapping operation. If you even compare these two APIs side-by-side, you'll see that the `SuggestionList` is not introducing anything meaningful by controlling the way the objects are mapped:

```jsx
<SuggestionList 
  suggestions={[{label: 'suh'}, {label: 'dude'}]} 
  renderListItem={(suggestion) => (
    <div>{suggestion.label}</div>
  )} 
/>
```

```jsx
<SuggestionList>
  {[{label: 'suh'}, {label: 'dude'}].map((suggestion) => (
    <div>{suggestion.label}</div>
  )} 
</SuggestionList>
```

The caller really has ultimate control, and there is no benefit in the `SuggestionList` pretending to have control.

Also, I don't think I ever explicitly made a suggestion, but what I implied by suggesting that we don't need this API is that the `SuggestionList` should stick to just accepting the `children` prop and exporting the `SuggestionListItem`.

**Comment by marioharper** [comment]
@GeorgeTaveras1231 Yeah I see what you mean, and this would keep in line with our `RadioGroup` and `Checkgroup` API's we have. It would just mean you would have to map over the array twice. Once on the consumers side, and again on our side in order to clone each child and pass in any props from the `SuggestionList`. The API proposed would mean we would give them a hook into that single `map` we do on our side. That's really the only benefit I could see.

**Comment by marioharper** [comment]
But this would be negligible if the result set isn't very large or has a default max limit set.

**Comment by GeorgeTaveras1231** [comment]
@marioharper I agree, it is more than likely a negligible difference in performance. I also agree this would be more in line with our other APIs. If we are inconsistent about it, we will confuse people lol.

**Comment by marioharper** [comment]
I vote to change it since it'll match our other API's

**Comment by GeorgeTaveras1231** [comment]
@hsbacot I will assign you this issue as you were the original implementer. @marioharper and I seem to be in agreement but just want to get your confirmation before moving on and creating a ticket to make this change.

**Comment by hsbacot** [comment]
The issue with having it match the other API's is that many use cases support having more than just a list of items. Each spec I've seen that would depend on this has a different footer or potentially multiple lists of suggestions based on one input.

I agree that there could be a better way to name the properties of a suggestion, however adding click handlers to each child isn't an optimal solution since they would be triggered when other elements, i.e. a footer or a label, are clicked. There are some ways around this to check the type of each child before applying a click handler while mapping through the list, though that might be a little more magical to a consumer.

The current API does not prevent the consumer from passing the items through via children. When given an API response of:
```jsx
{
  places: [ 'Place One', 'Place Two', 'Place Three' ]
}
```
The current implementation could handle the dynamic response in a few ways that should meet all needs.

```jsx
const suggestions = places.map(place => { description: place, id: place })
<SuggestionList suggestions={suggestions} onClick={() => alert('suggestion clicked')}>
  <Footer />
<SuggestionsList />
```
would be the same as
```jsx
const suggestions = places.map(place => (
  <SuggestionListItem id={place} onClick={() => alert('suggestion clicked')}>{place}</SuggestionListItem>)
)
<SuggestionList>
  {suggestions}
  <Footer />
</SuggestionList>
```
which is in line with they other API's, with the exception of adding a click handler to every child element (which isn't ideal in this case).

**Comment by GeorgeTaveras1231** [comment]
@hsbacot I think that the last example is not bad, especially if the expectation is that **some** children will need a custom handler(s) (Such as the `Footer` component). I think it is ok to require consumers to pass the handlers they need to **all** the child elements (as the example that maps to create `SuggestionListItem`s.

Anyways though, if you feel strongly about this, I'm ok with giving this up. 😄 

@smanwaring @marioharper any final thoughts on this? I'm cool with closing this if you guys are also cool with this decision.

**Comment by marioharper** [comment]
I like the second example shown
```jsx
<SuggestionList>
  {suggestions}
  <Footer />
</SuggestionList>
```
I almost feel this should be the default documentation. And the fact you can add `onClick` and `suggestions` to `SuggestionList` should be secondary. 

My use case would be to return the entire `suggestion` object back onClick. With the first example there is no way to do this, correct? Since it only returns an event object?

**Comment by GeorgeTaveras1231** [comment]
@marioharper what do you mean by ?
> My use case would be to return the entire suggestion object back onClick

**Comment by marioharper** [comment]
@GeorgeTaveras1231 

```jsx
const suggestions = places.map(place => (
  // returning place onClick
  <SuggestionListItem id={place.id} onClick={() => place}>{place.name}</SuggestionListItem>)
)
<SuggestionList>
  {suggestions}
  <Footer />
</SuggestionList>
```

**Comment by hsbacot** [comment]
Ok, I can update the docs to have this as well. @GeorgeTaveras1231 
@marioharper yes, that's correct.

**Comment by GeorgeTaveras1231** [comment]
Cool, @hsbacot feel free to close this once the update is made.
